### ISSUE TITLE
How to re-ingest data into a clone?

### ISSUE THREAD

**Issue by bpoldrack** [psychoinformatics-de/knowledge-base#12]
Origin: Datalad Office Hour Channel + subsequent https://github.com/datalad/datalad/issues/7348

Originally, the user's question was about how to transfer a dataset using an external hard drive as transportation. It was understood, that the straight-forward solution is to make a clone from the source system on the hard drive, get the data and then clone from the hard drive at the target system (and get the data).

However, the user had a copy of the data without a repository already (so,  snapshot essentially) and asked whether there was a way to save another round of expensive data transfer by only getting a plain clone without the data and then re-ingest the already existing data. It was suggested to unlock the files in the target clone, copy the data over and lock again in order to have git-annex move the content to the object tree.
The `directory` special remote was also pointed out as something worth looking at but with
a) no clear immediate idea how exactly to do it and whether that would work and
b) pointing out, that this approach would leave a trace of the special remote behind (which would immediately become useless)

The suggested method seemed to work initially, but slightly different workflow led to confusing state and behavior in the target clone (see https://github.com/datalad/datalad/issues/7348) involving adjusted branches, likely because the source system is a Windows machine, while the target system is Linux.

TODO (not necessarily to be performed in this order)
- [x] Inform OP/Add reference to this issue at origin
- [ ] Clarifying Qs asked or not needed
- [ ] Nature of the issue is understood
- [ ] Inform OP about resolution

**Comment by mih** [comment]
The situation is not entirely clear to me. Can you confirm that the following is the problem:

- the user wants to "move/duplicate" an entire DataLad dataset from a Windows machine to a Linux machine
- for a particular reason, they have to use an external drive for a "sneaker-net" transfer
- they want to avoid transferring file content from the Windows machine to the external drive, because it would only be in a "DataLad-specific" format, and they already have a copy of -- what they believe -- is the tracked file content

?

**Comment by bpoldrack** [comment]
@mih 

More or less, yes.

- the user wants a clone from a Windows machine on a linux machine
- for a particular reason, they have to use an external drive for a "sneaker-net" transfer
- it was realized that it could have been easy by making a clone on that drive and then from that drive
- there was, however, already a plain non-datalad/git/annex copy made of what they believe is the tracked content
- they want to avoid repeating the data transfer via datalad/annex and figure how it could work by only "adding the git part" this way and utilize the already transfered data

**Comment by mih** [comment]
The following log of a shell sessions shows the principle setup and an approach that avoids making another copy when one already has one

```sh
#
# Demo setup
# 
# Create a dataset on a FAT32 drive (simulating conditions similar to windows)
❯ datalad create /media/mih/Samsung_T5/tmp/onwindows
[INFO   ] Detected a filesystem without fifo support. 
[INFO   ] Disabling ssh connection caching. 
[INFO   ] Detected a crippled filesystem. 
[INFO   ] Entering an adjusted branch where files are unlocked as this filesystem does not support locked files. 
[INFO   ] Switched to branch 'adjusted/master(unlocked)' 
create(ok): /media/mih/Samsung_T5/tmp/onwindows (dataset)

# create a 10MB demo file (any number of files and sizes would not make a difference re this demo)
❯ datalad run -d /media/mih/Samsung_T5/tmp/onwindows 'dd if=/dev/urandom bs=1M count=10 of={dspath}/big.dat'
[INFO   ] == Command start (output follows) ===== 
10+0 records in
10+0 records out
10485760 bytes (10 MB, 10 MiB) copied, 0.0254713 s, 412 MB/s
[INFO   ] == Command exit (modification check follows) ===== 
run(ok): /media/mih/Samsung_T5/tmp/onwindows (dataset) [dd if=/dev/urandom bs=1M count=10 of=/me...]
add(ok): big.dat (file)                                                                                        
save(ok): . (dataset)                                                                                          

# the file is a regular file, no symlink
❯ ls -l /media/mih/Samsung_T5/tmp/onwindows
.rwxr-xr-x 10M mih mih 14 Apr 11:19  big.dat

#
# Create a clone on the target system
#
# This is equivalent to taken care of just the "Git part" of a dataset. It is not necessary to
# create a straight-up clone on the target system directly, but it will suffice for this demo.
#
❯ datalad clone /media/mih/Samsung_T5/tmp/onwindows /tmp/onlinux
install(ok): /tmp/onlinux (dataset) 

# when cloning an adjust-mode repository directly from a checkout, datalad is not able to
# establish a proper clone. It will look like a dataset in adjust-mode, but it is not.
# Putting it into adjusted-mode via `git-annex adjust --unlock` also does not yield a working
# dataset.
# We simply remove the adjustment instead.

# kill the adjust branch and checkout master/main
❯ git -C /tmp/onlinux checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
❯ git -C /tmp/onlinux branch -D adjusted/master\(unlocked\)
Deleted branch adjusted/master(unlocked) (was 817e68f).

# now we have a regular "normal" mode dataset
❯ ls -l /tmp/onlinux
lrwxrwxrwx 130 mih mih 14 Apr 11:20  big.dat -> .git/annex/objects/k8/q3/MD5E-s10485760--48ecd5908c26ce5bcc1c2b3901a5120f.dat/MD5E-s10485760--48ecd5908c26ce5bcc1c2b3901a5120f.dat

# it is clean
❯ datalad status -d /tmp/onlinux
nothing to save, working tree clean

# here is the current commit (pay attention to the gitsha)
❯ git -C /tmp/onlinux log -1 --pretty=oneline | cat
11008c27a197fd4fe5e61a00d71a594ca1a03ef5 [DATALAD RUNCMD] dd if=/dev/urandom bs=1M count=10 of=/me...

# now we can simply copy the data snapshots into/over the worktree
❯ cp --remove-destination /media/mih/Samsung_T5/tmp/onwindows/big.dat /tmp/onlinux

# `status` will detect the "change", because we replaced the symlink provided by git-annex
# with the actual data file
❯ datalad status -d /tmp/onlinux
 modified: big.dat (symlink)
❯ ls -l /tmp/onlinux
.rwxr-xr-x 10M mih mih 14 Apr 11:00 big.dat

# 
# Here is the key
# 
# all that is needed to "ingest" the file content is a `save`
# if everything is as expected, i.e. the data copied is actually the data that is expected/tracked
# this will not result in a new dataset state (as seen below) -- content is simply moved into the
# annex.
# however, if data was corrupted this step would notice it and produce a recorded checke
❯ datalad save -d /tmp/onlinux -m "Reingest data after move from ..."
add(ok): big.dat (file)                                                                                        
action summary:                                                                                                
  add (ok: 1)
  save (notneeded: 1)
❯ datalad status -d /tmp/onlinux
nothing to save, working tree clean
❯ git -C /tmp/onlinux log -1 --pretty=oneline | cat
11008c27a197fd4fe5e61a00d71a594ca1a03ef5 [DATALAD RUNCMD] dd if=/dev/urandom bs=1M count=10 of=/me...
```

**Comment by bpoldrack** [comment]
FTR: What is described above is pretty much what I quickly tried and suggested in office hour chat. What I did not consider and failed to realize was the cloning from an adjusted branch repo. Which looks like what led to the subsequent https://github.com/datalad/datalad/issues/7348

**Comment by mih** [comment]
I believe this should turn into two KBIs

- How can I clone a dataset that exists in the form of an adjusted mode checkout?
  - It remains to be tested whether cloning adjusted-mode -> adjusted-mode works any better than what we see here
- How can I re-ingest file content in a dataset clone?

**Comment by adswa** [comment]
I have created two KBIs based on your demo script - thanks much for the thorough script and comments!

**Comment by mih** [comment]
This has been resolved by #37 and #38
